devices/wholeBodyDynamics/WholeBodyDynamicsDevice.cpp

```diff
@@ -303,78 +303,257 @@ bool WholeBodyDynamicsDevice::openEstimator(os::Searchable& config)
+            indexesToDiscard.push_back(i);
         }
         else
         {
-            defaultContactFramesIdx.push_back(idx);
+            contactFramesIdx.push_back(idx);
         }
     }
 
+    sort(indexesToDiscard.begin(), indexesToDiscard.end());
+    //remove the variables that correspond to invalid frames
+    for(int i=indexesToDiscard.size()-1; i>=0; i-- )
+    {
+        contactFramesNames.erase(contactFramesNames.begin() + indexesToDiscard[i]);
+        unknownExtWrench.erase(unknownExtWrench.begin() + indexesToDiscard[i]);
+        nrUnknownsInExtWrench.erase(nrUnknownsInExtWrench.begin() + indexesToDiscard[i]);
+    }
+
+    // The frames from in 1 submodel should be no more than 1 if the type is FULL_WRENCH, no more than 2 if the type is PURE_FORCE, and no more than 6 if the type is PURE_FORCE_WITH_KNOWN_DIRECTION.
+    // Refer to section 3 of the theory paper https://www.researchgate.net/publication/236152161_Contact_Force_Estimations_Using_Tactile_Sensors_and_Force_Torque_Sensors
+
     // For each submodel, we find the first suitable contact frame
     // This is a n^2 algorithm, but given that is just used in
     // configuration it should be ok
     size_t nrOfSubModels = estimator.submodels().getNrOfSubModels();
 
-    // We indicate with FRAME_INVALID_INDEX the fact that we still don't have a default contact for the given submodel
-    subModelIndex2DefaultContact.resize(nrOfSubModels,iDynTree::FRAME_INVALID_INDEX);
+    // We indicate with FRAME_INVALID_INDEX the fact that we still don't have a default/override contact for the given submodel
+    contactFramesIdxValidForSubModel.resize(contactFramesIdx.size(),iDynTree::FRAME_INVALID_INDEX);
+    //create a vector to store the submodel indexes that will have contact points
+    std::vector<size_t> validSubModelsIndexes;
+    validSubModelsIndexes.resize(contactFramesIdx.size(),iDynTree::FRAME_INVALID_INDEX);
+
+    nrUnknownsInSubModel.resize(nrOfSubModels,0);
 
-    for(size_t i=0; i < defaultContactFramesIdx.size(); i++)
+    for(size_t i=0; i < contactFramesIdx.size(); i++)
     {
-        size_t subModelIdx = estimator.submodels().getSubModelOfFrame(estimator.model(),defaultContactFramesIdx[i]);
+        size_t subModelIdx = estimator.submodels().getSubModelOfFrame(estimator.model(),contactFramesIdx[i]);
 
-        // If the subModel of the frame still does not have a default contact, we add it
-        if( subModelIndex2DefaultContact[subModelIdx] == iDynTree::FRAME_INVALID_INDEX )
+        //check if adding the current contact to this subModel will increase the no of variable to more than 6
+        if(nrUnknownsInSubModel[subModelIdx]+nrUnknownsInExtWrench[i] > 6)
         {
-            subModelIndex2DefaultContact[subModelIdx] = defaultContactFramesIdx[i];
+            yWarning() << "subModel no " << subModelIdx << " has the maximum number of variables (6). The frame " << contactFramesNames[i] << " will not be added to it.";
```

Comment:
What is the rationale in not exiting as early as possible if the configuration file has an error?

Reply:
To be honest I haven't thought about it when I put it. I  tried to keep similar logic to what was there before when someone puts an invalid frame name in `defaultContactFrames` (or a frame not existing in the model), the device doesn't exist only throws a warning.
https://github.com/robotology/whole-body-estimators/blob/22302eef59ad2f0b900a662490850206a6aea909/devices/wholeBodyDynamics/WholeBodyDynamicsDevice.cpp#L332

I think it better to keep it because when I was testing the estimation. I wanted to change one contact from `pureKnown` to `full` to see how the estimation quality will change, and I didn't want to bother checking and removing the rest of the contacts of the same submodel in the list. The device may proceed and remove them for me!